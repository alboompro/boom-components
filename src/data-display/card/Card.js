import React from "react";
import PropTypes from "prop-types";

import {
  CardStyle,
  HeadStyle,
  BodyStyle,
  TitleStyle,
  ExtraStyle,
  ActionStyle
} from "./styles";

const Card = ({ children, title, style, cover, extra, actions, ...props }) => {
  const StyleProps = {
    bordered: props.bordered,
    hoverable: props.hoverable,
    ...props
  };

  const HeadProps = {
    size: props.size
  };

  const BodyProps = {
    size: props.size,
    cover: cover
  };

  let head = React.ReactNode; // Se TS seria head: React.ReactNode (anotação de tipo)
  if (title || extra) {
    head = (
      <HeadStyle {...HeadProps}>
        {title && <TitleStyle {...HeadProps}>{title}</TitleStyle>}
        {extra && <ExtraStyle {...HeadProps}>{extra}</ExtraStyle>}
      </HeadStyle>
    );
  }

  const coverDOM = cover ? { cover } : null;

  const body = <BodyStyle {...BodyProps}>{children}</BodyStyle>;

  let getAction = actions => {
    const actionList = actions.map((action, index) => (
      <li style={{ width: `${100 / actions.length}%` }} key={index}>
        <span>{action}</span>
      </li>
    ));
    return actionList;
  };

  const actionDOM =
    actions && actions.length ? (
      <ActionStyle>{getAction(actions)}</ActionStyle>
    ) : null;

  // const actionDOM =

  return (
    <CardStyle {...StyleProps} style={{ ...style }}>
      {head}
      {cover ? cover : ""}
      {body}
      {actionDOM}
    </CardStyle>
  );
};

Card.propTypes = {
  /** actions list which shows at the bottom of the card */
  actions: PropTypes.arrayOf(PropTypes.node),
  /** whether card will have a border around of it */
  bordered: PropTypes.bool,
  /** lift up card when hovering */
  hoverable: PropTypes.bool,
  /** show a skeleton instead of content */
  loading: PropTypes.bool,
  /** size of the card */
  size: PropTypes.oneOf(["default", "small"]),
  /** card cover */
  cover: PropTypes.node,
  /** card title */
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  /** content to render in the top-right corner of the card */
  extra: PropTypes.node
};

Card.defaultProps = {
  actions: null,
  bordered: true,
  hoverable: false,
  loading: false,
  size: "default",
  cover: null,
  extra: null,
  title: null
};

export default Card;
