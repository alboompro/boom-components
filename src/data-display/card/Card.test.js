import React from "react";
import { mount } from "enzyme";
import Card from "./Card";
import {
  CardStyle,
  HeadStyle,
  BodyStyle,
  TitleStyle,
  ExtraStyle,
  ActionStyle
} from "./styles";

describe("Card", () => {
  test("card props", () => {
    const wrapper = mount(
      <Card
        style={{ width: 240 }}
        cover={
          <img
            alt="image example"
            src="https://images.pexels.com/photos/1252983/pexels-photo-1252983.jpeg?cs=srgb&dl=pexels-lisa-1252983.jpg&fm=jpg"
          />
        }
        title="Card test with Cover"
        size="small"
        hoverable
      >
        <p>children</p>
      </Card>
    );

    const CardStyle = wrapper.find(CardStyle).props();
    const HeadStyle = wrapper.find(HeadStyle).props();
    const BodyStyle = wrapper.find(BodyStyle).props();
    const TitleStyle = wrapper.find(TitleStyle).props();
    const ExtraStyle = wrapper.find(ExtraStyle).props();
    const ActionStyle = wrapper.find(ActionStyle).props();

    expect(HeadStyle.props().size).toEqual("small");
    expect(TitleStyle.props().title).toEqual("Card test with Cover");
    expect(CardStyle.props().hoverable).toEqual(true);
    expect(CardStyle.props().cover).toEqual(
      <img
        alt="image example"
        src="https://images.pexels.com/photos/1252983/pexels-photo-1252983.jpeg?cs=srgb&dl=pexels-lisa-1252983.jpg&fm=jpg"
      />
    );
    expect(CardStyle.props().style).toEqual({ width: 240 });
    expect(BodyStyle.props().children).toEqual(<p>children</p>);
  });
});
