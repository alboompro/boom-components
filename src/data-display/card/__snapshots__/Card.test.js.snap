// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Props - Snapshot could create a card with Meta 1`] = `
<styled.div
  bordered={true}
  hoverable={false}
  loading={false}
  size="small"
  style={
    Object {
      "width": 300,
    }
  }
>
  <div>
    <styled.div
      size="small"
    >
      <styled.div
        size="small"
      >
        Card Test with Meta
      </styled.div>
    </styled.div>
  </div>
  <img
    alt="image example"
    src="htps://images.pexels.com/photos/1252983/pexels-photo-1252983.jpeg?cs=srgb&dl=pexels-lisa-1252983.jpg&fm=jpg"
  />
  <styled.div
    cover={
      <img
        alt="image example"
        src="htps://images.pexels.com/photos/1252983/pexels-photo-1252983.jpeg?cs=srgb&dl=pexels-lisa-1252983.jpg&fm=jpg"
      />
    }
    size="small"
  >
    <Meta
      avatar={
        <img
          alt="image example"
          src="https://images.unsplash.com/photo-1522075469751-3a6694fb2f61?ixid=MnwxMjA3fDB8MHxzZWFyY2h8Nnx8cHJvZmlsZSUyMHBpY3xlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60"
        />
      }
      description="Test meta-description"
      title="Test meta-title"
    />
  </styled.div>
  <styled.ul>
    <li
      key="0"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <span>
        <Button
          className={null}
          disabled={false}
          href={null}
          htmlType="button"
          icon={null}
          iconStyle={Object {}}
          loading={true}
          loadingColor="#333"
          loadingSize={18}
          onClick={[Function]}
          outlined={false}
          rounded={false}
          size="default"
          stretched={false}
          style={Object {}}
          target={null}
          textStyle={Object {}}
          type="main"
        >
          Waiting
        </Button>
      </span>
    </li>
  </styled.ul>
</styled.div>
`;

exports[`Props - Snapshot could create a card with cover 1`] = `
<styled.div
  bordered={true}
  hoverable={true}
  loading={false}
  size="small"
  style={
    Object {
      "width": 240,
    }
  }
>
  <div>
    <styled.div
      size="small"
    >
      <styled.div
        size="small"
      >
        Card Test with Cover
      </styled.div>
    </styled.div>
  </div>
  <img
    alt="image example"
    src="htps://images.pexels.com/photos/1252983/pexels-photo-1252983.jpeg?cs=srgb&dl=pexels-lisa-1252983.jpg&fm=jpg"
  />
  <styled.div
    cover={
      <img
        alt="image example"
        src="htps://images.pexels.com/photos/1252983/pexels-photo-1252983.jpeg?cs=srgb&dl=pexels-lisa-1252983.jpg&fm=jpg"
      />
    }
    size="small"
  >
    <p>
      child component text
    </p>
  </styled.div>
</styled.div>
`;
